redhat-appstudio/service-provider-integration-operator
# Refresh token based on token label

### What does this PR do?
If a SPIAccessToken has the label "spi.appstudio.redhat.com/token-refresh" the controller will try to refresh the token. This is a temporary solution as users should not have to "instruct" the controller. However, the PR serves the purpose of implementing token refresh logic without scheduling the refresh by itself.

### Screenshot/screencast of this PR



### What issues does this PR fix or reference?

implements https://issues.redhat.com/browse/SVPI-189

### How to test this PR?

1. Deploy with `SPIO_IMG=quay.io/pbaran/spio:refresh-token`
2. Edit the service provider configuration and add GitLab `clientId` and `clientSecret`. Restart spis/spio.
3. Create GitLab SPIAccessTokenBinding such as `samples/binding-write-repo-gitlab.yaml`
4. Go through the OAuth flow and check and save somewhere the contents of the Kubernetes secret that the binding references
5. Edit the SPIAccessToken that the binding from step 3. links to, by adding a label `spi.appstudio.redhat.com/refresh-token: "true"`
6. Check after a short time that the added label from the previous step is removed from the SPIAccessToken
7. As in the step 4, check the referenced secret's contents and compare them with the previous. The access tokens should be different.
8. Logs of operator should contain `"msg":"token refreshed successfully"`
